src/EditorFeatures/Core/Remote/SolutionChecksumUpdater.cs

```diff
@@ -21,115 +19,148 @@ namespace Microsoft.CodeAnalysis.Remote
         {
-            base.Shutdown();
-
-            // stop listening workspace change event
-            _workspace.WorkspaceChanged -= OnWorkspaceChanged;
-
-            CancelAndDispose(_globalOperationCancellationSource);
+            // create a new token to control all the work we need to do while we're current unpaused.
+            var nextToken = _pauseCancellationSeries.CreateNext();
+            _currentWorkToken = nextToken;
+            // Start synchronizing again.
+            _synchronizeWorkspaceQueue.AddWork(nextToken);
         }
 
-        private void OnWorkspaceChanged(object sender, WorkspaceChangeEventArgs e)
+        private void OnWorkspaceChanged(object? sender, WorkspaceChangeEventArgs e)
         {
-            if (e.Kind == WorkspaceChangeKind.DocumentChanged)
-            {
-                PushTextChanges(e.OldSolution.GetDocument(e.DocumentId), e.NewSolution.GetDocument(e.DocumentId));
-            }
-
-            // record that we are busy
-            UpdateLastAccessTime();
-
-            EnqueueChecksumUpdate();
-        }
+            // Check if we're currently paused.  If so ignore this notification.  We don't want to any work in response
+            // to whatever the workspace is doing.
+            var workToken = _currentWorkToken;
+            if (workToken.IsCancellationRequested)
+                return;
 
-        private void EnqueueChecksumUpdate()
-        {
-            // event will raised sequencially. no concurrency on this handler
-            if (_workQueue.TryPeek(out _))
+            if (e.Kind == WorkspaceChangeKind.DocumentChanged)
             {
-                return;
+                _textChangeQueue.AddWork((e.OldSolution.GetDocument(e.DocumentId), e.NewSolution.GetDocument(e.DocumentId)));
             }
 
-            _workQueue.Enqueue(Listener.BeginAsyncOperation(nameof(SolutionChecksumUpdater)));
+            _synchronizeWorkspaceQueue.AddWork(workToken);
         }
 
-        private async Task SynchronizePrimaryWorkspaceAsync(CancellationToken cancellationToken)
+        private async ValueTask SynchronizePrimaryWorkspaceAsync(
+            ImmutableSegmentedList<CancellationToken> cancellationTokens,
+            CancellationToken disposalToken)
         {
             var solution = _workspace.CurrentSolution;
             if (solution.BranchId != _workspace.PrimaryBranchId)
-            {
                 return;
-            }
+
+            // Check if we've been asked to cancel because we've been paused.  If so, don't bother proceeding.
+            var uncancelledTokens = cancellationTokens.Where(ct => !ct.IsCancellationRequested).ToList();
+            if (uncancelledTokens.Count == 0)
+                return;
+
+            uncancelledTokens.Add(disposalToken);
+
+            // Create a token that will fire if we are disposed or if we get paused.
+            using var source = CancellationTokenSource.CreateLinkedTokenSource(uncancelledTokens.ToArray());
```

Comment:
At this point, should uncancelledTokens contain exactly one entry? Because if all of the tokens came from the CancellationSeries, it seems only one should ever be non-cancelled?

Reply:
It could contain either one or zero (zero if they were all canceled).  One if all but hte last were canceled.  I can change this to SingleOrNull due to that.

Reply:
Changed.  Thanks!